### ISSUE TITLE
no-unused-vars reports exported variables as unused in node environment

### ISSUE THREAD

**Issue by ilyavolodin** [eslint/eslint#3916]
ESLint v1.5.0
The following code:
```js
/*eslint-env node */
/*eslint no-unused-vars: 2*/

var foo = 9;

/* exported foo */
```
reports the following:
```sh
4:5 - "foo" is defined but never used (no-unused-vars)
```

expected behavior is to not have any errors. This is due to `globalReturn` flag being on for node environment. I assume the same happens for commonjs environment as well.

**Comment by mysticatea** [comment]
This behavior seems to be intented: https://github.com/eslint/eslint/blob/560d7086051eff720fb63a877f6f8b95e9235316/tests/lib/eslint.js#L1273-L1299

**Comment by ilyavolodin** [comment]
Hmm... Ok, I think that makes sense for Node environment. But the same would be happening for CommonJS environment, and I'm not 100% sure it makes sense there. Anyone has an opinion? @eslint/eslint-team

**Comment by nzakas** [comment]
I think marking something as exported should count as being used regardless of scoping.

**Comment by michaelficarra** [comment]
@nzakas The more likely scenario is that either the environment is wrong or they didn't mean to have that comment. These are conflicting configurations.

**Comment by nzakas** [comment]
That's a very good point. @ilyavolodin what do you think?

**Comment by ilyavolodin** [comment]
I agree with @michaelficarra statement in regards to Node. However, as far as I remember, we added CommonJS environment for browserified javascript. I can see how you might want part of the script to be CommonJS (as in module.exports) but also want to assign something to a global variable. I know it's a stretch. So I'm fine if we don't implement this, but in that case, I think we need clear explanation in documentation that `exported` is not supported in Node and CommonJS or when modules are on.

**Comment by nzakas** [comment]
Okay, let's go with a documentation update for now.

**Comment by STRML** [comment]
I'm finding that this would be useful for a pattern I often use as a form of self-documentation:

`let FooComponent = module.exports = someConstructor({...`

Unfortunately, `no-unused-vars` always marks `FooComponent` as unused because it is not referenced elsewhere in the file.

Setting `/*eslint no-unused-vars:0*/` shuts off var checking for the whole file. For now I've been prepending an underscore (to use `varsIgnorePattern`) but this is pretty annoying. Is there any way to support this syntax?

**Comment by nzakas** [comment]
Working on this.

**Comment by revelt** [comment]
Hi all, I came here researching why JS Standard was flagging up my exported variables as unused. It's strange, since the pattern is quite common and legit, it's not forbidden to put `module.exports = function(...`. I don't see why variables inside of that function should be flagged as unused.

**Comment by platinumazure** [comment]
@revelt This issue is about using an `/* exported */ ` comment to tell ESLint that a variable is used somewhere else, rather than assigning to `module.exports` or anything like that. Please open a new issue or stop by our [Gitter chat](https://gitter.im/eslint/eslint).
